### PR TITLE
feat: adds search to changeling evolution menu

### PR DESCRIPTION
**Pull request by Gaxeer**
## What Does This PR Do
Adds search by name and description to changeling evolution menu

## Why It's Good For The Game
QoL

## Images of changes
<details>
 <summary>Demo video</summary>
 

https://github.com/ParadiseSS13/Paradise/assets/44334376/771582b3-eb68-428e-965b-2840611fb0a2

</details>

## Testing
Tested search ingame: searched, evolved, reevolved - seems fine

## Changelog
:cl:
add: Adds search by name and description to changeling evolution menu
/:cl:

### CODE REVIEW

**Review by S34NW — COMMENTED** [review]
Can you merge `upstream/master` and recompile TGUI to deconflict, please?

**Comment by Gaxeer** [comment]
> Can you merge `upstream/master` and recompile TGUI to deconflict, please?

Fixed. Also did a small refactor to make categories more flexible and to make search work properly.

File: code/modules/antagonists/changeling/changeling_power.dm [hunk 1]
```diff
@@ -5,14 +5,16 @@
 /// Denotes that this power can only be obtained by purchasing it.
 #define CHANGELING_PURCHASABLE_POWER	2
 /// Denotes that this power can not be obtained normally. Primarily used for base types such as [/datum/action/changeling/weapon].
-#define CHANGELING_UNOBTAINABLE_POWER	3
+#define  CHANGELING_UNOBTAINABLE_POWER	3
```
Comment (Henri215) [inline code/modules/antagonists/changeling/changeling_power.dm:8]:
```suggestion
#define CHANGELING_UNOBTAINABLE_POWER	3
```

File: code/modules/antagonists/changeling/evolution_menu.dm [hunk 2]
```diff
@@ -110,3 +86,55 @@
...
+	PRIVATE_PROC(TRUE)
+
+	var/list/abilities_by_category_name = get_abilities_grouped_by_category_name()
+	if (!length(abilities_by_category_name))
+		return list()
+
+	var/list/sorted_ability_categories = sortTim(subtypesof(/datum/changeling_power_category), GLOBAL_PROC_REF(cmp_changeling_power_category_asc))
+
+	var/list/sorted_ability_tabs = list()
+	for (var/datum/changeling_power_category/category as anything in sorted_ability_categories)
```
Comment (Henri215) [inline code/modules/antagonists/changeling/evolution_menu.dm:108]:
```suggestion
	if(!length(abilities_by_category_name))
		return list()

	var/list/sorted_ability_categories = sortTim(subtypesof(/datum/changeling_power_category), GLOBAL_PROC_REF(cmp_changeling_power_category_asc))

	var/list/sorted_ability_tabs = list()
	for(var/datum/changeling_power_category/category as anything in sorted_ability_categories)
```

File: code/modules/antagonists/changeling/evolution_menu.dm [hunk 3]
```diff
@@ -110,3 +86,55 @@
...
+	for(var/power_path in cling.purchaseable_powers)
+		var/datum/action/changeling/changeling_ability = power_path
+
+		if(!changeling_ability.category)
+			stack_trace("Cling power [changeling_ability], [changeling_ability.type] had no assigned category!")
+			continue
+
+		var/category_name = initial(changeling_ability.category.name)
+		var/list/abilities = abilities_by_category_name[category_name]
+		if (!islist(abilities))
```
Comment (Henri215) [inline code/modules/antagonists/changeling/evolution_menu.dm:129]:
```suggestion
		if(!islist(abilities))
```

**Review by S34NW — APPROVED** [review]

**Review by S34NW — APPROVED** [review]
Deconflict for merge

**Pull request merged**